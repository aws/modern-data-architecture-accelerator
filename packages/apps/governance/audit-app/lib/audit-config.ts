/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { MdaaAppConfigParser, MdaaAppConfigParserProps, MdaaBaseConfigContents } from '@aws-mdaa/app';
import { BucketInventoryProps } from '@aws-mdaa/audit-l3-construct';
import { MdaaRoleRef } from '@aws-mdaa/iam-role-helper';
import { Schema } from 'ajv';
import { Stack } from 'aws-cdk-lib';

import * as configSchema from './config-schema.json';

export interface AuditConfigContents extends MdaaBaseConfigContents {
  /**
   * List of refs to roles which will be provided read access to the audit logs
   */
  readRoles?: MdaaRoleRef[];
  /**
   * List of source accounts from which audit logs will be accepted
   */
  sourceAccounts?: string[];
  /**
   * List of source regions from which audit logs will be accepted
   */
  sourceRegions?: string[];
  /**
   * Specifies the S3 prefix where inventory data will be accepted on the audit bucket
   */
  inventoryPrefix?: string;
  /**
   * The list of expected inventories
   */
  inventories?: BucketInventoryProps[];
}

export class AuditConfigParser extends MdaaAppConfigParser<AuditConfigContents> {
  public readonly readRoleRefs: MdaaRoleRef[];
  public readonly sourceAccounts: string[];
  public readonly sourceRegions: string[];
  public readonly inventoryPrefix: string;
  public readonly inventories?: BucketInventoryProps[];

  constructor(stack: Stack, props: MdaaAppConfigParserProps) {
    super(stack, props, configSchema as Schema);

    this.readRoleRefs = this.configContents.readRoles ? this.configContents.readRoles : [];
    this.sourceAccounts = this.configContents.sourceAccounts ? this.configContents.sourceAccounts : [];
    this.sourceRegions = this.configContents.sourceRegions ? this.configContents.sourceRegions : [];
    this.inventoryPrefix = this.configContents.inventoryPrefix ? this.configContents.inventoryPrefix : 'inventory/';
    this.inventories = this.configContents.inventories;
  }
}
