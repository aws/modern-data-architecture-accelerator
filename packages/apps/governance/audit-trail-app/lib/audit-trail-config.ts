/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { CaefAppConfigParser, CaefAppConfigParserProps, CaefBaseConfigContents } from '@aws-caef/app';
import { AuditTrailProps } from '@aws-caef/audit-trail-l3-construct';
import { Schema } from 'ajv';
import { Stack } from 'aws-cdk-lib';
import * as configSchema from './config-schema.json';

export interface AuditTrailConfigContents extends CaefBaseConfigContents {
    readonly trail: AuditTrailProps
}

export class AuditTrailConfigParser extends CaefAppConfigParser<AuditTrailConfigContents> {

    public readonly trail: AuditTrailProps

    constructor( stack: Stack, props: CaefAppConfigParserProps ) {
        super( stack, props, configSchema as Schema )
        this.trail = this.configContents.trail
    }

}

