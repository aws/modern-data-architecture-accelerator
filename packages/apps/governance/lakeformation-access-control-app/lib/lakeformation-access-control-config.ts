/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { MdaaAppConfigParser, MdaaAppConfigParserProps, MdaaBaseConfigContents } from '@aws-mdaa/app';
import { MdaaRoleRef } from '@aws-mdaa/iam-role-helper';
import {
  LakeFormationAccessControlL3Construct,
  NamedGrantProps,
  NamedPrincipalProps,
  NamedResourceLinkProps,
  PermissionsConfig,
  ResourceLinkProps,
} from '@aws-mdaa/lakeformation-access-control-l3-construct';
import { Schema } from 'ajv';
import { Stack } from 'aws-cdk-lib';
import * as configSchema from './config-schema.json';

export interface PrincipalConfig {
  /**
   * Federated group name for the grant.
   */
  readonly federatedGroup?: string;
  /**
   * Federated user name for the grant.
   */
  readonly federatedUser?: string;
  /**
   * Name of the federation provider from the federationProviders section of the config
   */
  readonly federationProvider?: string;
  /**
   * Arn of an IAM principal for the grant.
   */
  readonly role?: MdaaRoleRef;
}

export interface GrantConfig {
  /**
   * Name of the existing Glue Database to perform the grant against.
   */
  readonly database: string;
  /**
   * Permissions to Grant on database.  Must be 'read', 'write', or 'super'. Defaults to 'read'.
   */
  readonly databasePermissions?: PermissionsConfig;
  /**
   * Array of strings representing Tables to perform the grant against.  Use a '*' for all tables.
   */
  readonly tables?: string[];
  /**
   * Permissions to Grant on tables.  Must be 'read', 'write', or 'super'. Defaults to 'read'.
   */
  readonly tablePermissions?: PermissionsConfig;
  /**
   * Array of strings representing principals to grant permissions to.  These must exist in the 'principals:' section.
   */
  readonly principals: string[];
}

export interface FederationProviderConfig {
  [key: string]: string;
}

export interface ResourceLinkConfig {
  /**
   * Name of the target database
   */
  readonly targetDatabase: string;
  /**
   * The account where the target database exists
   */
  readonly targetAccount?: string;
  /**
   * The account in which the resource link should be created.
   * If not specified, will default to the local account.
   */
  readonly fromAccount?: string;
  /**
   * Array of strings representing principals to grant resource link DESCRIBE permissions to.
   */
  readonly grantPrincipals?: string[];
}

export interface LakeFormationAccessControlConfigContents extends MdaaBaseConfigContents {
  /**
   * Array of Objects representing the grants
   */
  readonly grants: { [key: string]: GrantConfig };
  /**
   * Object representing the active directory users, groups, and federation provider used
   */
  readonly principals: { [key: string]: PrincipalConfig };
  /**
   * Maps of Federation providers used in the principals: section to ARNs in IAM
   */
  readonly federationProviders?: FederationProviderConfig;
  /**
   * Resource links which will be created in the local account.
   */
  readonly resourceLinks?: { [key: string]: ResourceLinkConfig };
}

export class LakeFormationAccessControlConfigParser extends MdaaAppConfigParser<LakeFormationAccessControlConfigContents> {
  public readonly grants: NamedGrantProps;
  public readonly resourceLinks?: NamedResourceLinkProps;

  constructor(stack: Stack, props: MdaaAppConfigParserProps) {
    super(stack, props, configSchema as Schema);

    const resourceLinkPropsEntries = Object.entries(this.configContents.resourceLinks || {}).map(resourceLinkEntry => {
      const configResourceLinkName = resourceLinkEntry[0];
      const configResourceLink = resourceLinkEntry[1];

      const principals: NamedPrincipalProps = this.resolvePrincipals(configResourceLink.grantPrincipals || []);

      const resourceLinkProps: ResourceLinkProps = {
        targetDatabase: configResourceLink.targetDatabase,
        targetAccount: configResourceLink.targetAccount,
        grantPrincipals: principals,
        fromAccount: configResourceLink.fromAccount,
      };

      return [configResourceLinkName, resourceLinkProps];
    });

    this.resourceLinks = Object.fromEntries(resourceLinkPropsEntries);

    this.grants = Object.fromEntries(
      Object.entries(this.configContents.grants).map(configGrantEntry => {
        const configGrantName = configGrantEntry[0];
        const configGrant = configGrantEntry[1];
        const principals = this.resolvePrincipals(configGrant.principals);
        return [
          configGrantName,
          {
            ...configGrant,
            databasePermissions:
              LakeFormationAccessControlL3Construct.DATABASE_PERMISSIONS_MAP[configGrant.databasePermissions || 'read'],
            tablePermissions:
              LakeFormationAccessControlL3Construct.TABLE_PERMISSIONS_MAP[configGrant.tablePermissions || 'read'],
            principals: principals,
          },
        ];
      }),
    );
  }
  private resolvePrincipals(principals: string[]) {
    const resolvedPrincipals: NamedPrincipalProps = Object.fromEntries(
      principals.map(configPrincipalName => {
        const configPrincipal = this.configContents.principals[configPrincipalName];
        const federationProviderArn =
          configPrincipal.federationProvider && this.configContents.federationProviders
            ? this.configContents.federationProviders[configPrincipal.federationProvider]
            : undefined;
        if (configPrincipal.federationProvider && !federationProviderArn) {
          throw new Error(`Failed to resolve federation provider in config: ${configPrincipal.federationProvider}`);
        }
        return [
          configPrincipalName,
          {
            ...configPrincipal,
            federationProviderArn: federationProviderArn,
          },
        ];
      }),
    );
    return resolvedPrincipals;
  }
}
