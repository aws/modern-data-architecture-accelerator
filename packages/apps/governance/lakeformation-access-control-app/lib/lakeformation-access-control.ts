/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { MdaaAppConfigParserProps, MdaaCdkApp } from '@aws-mdaa/app';
import {
  LakeFormationAccessControlL3Construct,
  LakeFormationAccessControlL3ConstructProps,
} from '@aws-mdaa/lakeformation-access-control-l3-construct';
import { MdaaL3ConstructProps } from '@aws-mdaa/l3-construct';
import { AppProps, Stack } from 'aws-cdk-lib';
import { LakeFormationAccessControlConfigParser } from './lakeformation-access-control-config';

export class LakeFormationCdkApp extends MdaaCdkApp {
  constructor(props: AppProps = {}) {
    super(props, MdaaCdkApp.parsePackageJson(`${__dirname}/../package.json`));
  }
  protected subGenerateResources(
    stack: Stack,
    l3ConstructProps: MdaaL3ConstructProps,
    parserProps: MdaaAppConfigParserProps,
  ) {
    const appConfig = new LakeFormationAccessControlConfigParser(stack, parserProps);
    const constructProps: LakeFormationAccessControlL3ConstructProps = {
      ...{
        resourceLinks: appConfig.resourceLinks,
        grants: appConfig.grants,
      },
      ...l3ConstructProps,
    };

    new LakeFormationAccessControlL3Construct(stack, 'access', constructProps);
    return [stack];
  }
}
