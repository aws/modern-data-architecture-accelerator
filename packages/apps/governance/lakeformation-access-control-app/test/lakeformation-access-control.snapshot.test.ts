/*!
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import { describe } from '@jest/globals';
import { snapShotTest, snapShotTestApp, Create } from '@aws-mdaa/testing';
import { LakeFormationCdkApp } from '../lib/lakeformation-access-control';
import * as path from 'path';

describe('lakeformation-access-control Snapshot Tests', () => {
  beforeAll(() => {
    expect.addSnapshotSerializer({
      test: (val: unknown) => typeof val === 'string' && val.includes('[CONFIG:') && val.includes('test-config.yaml]'),
      print: (val: unknown) => {
        const stringVal = val as string;
        return `"${stringVal.replace(/\[CONFIG:[^[\]]*test-config\.yaml\]/, '[CONFIG:test-config.yaml]')}"`;
      },
    });
  });
  snapShotTest(
    'Lakeformation Stack',
    Create.stackProvider(
      'LakeformationStackMain',
      (_, context) => {
        const region = process.env.CDK_DEFAULT_REGION || 'us-east-1';

        const moduleApp = new LakeFormationCdkApp({
          context: {
            ...context,
            module_configs: path.join(__dirname, 'test-config.yaml'),
            additional_accounts: 'xxxxxxxxxxxxx',
            additional_stacks: [
              {
                account: 'xxxxxxxxxxxxx',
                region: region,
              },
            ],
          },
        });
        return moduleApp.generateStack();
      },
      {
        module_name: 'test-lakeformation-main',
        org: 'test-org',
        env: 'test-env',
        domain: 'test-domain',
      },
    ),
  );

  snapShotTestApp(
    'Lakeformation App',
    Create.appProvider(
      context => {
        const region = process.env.CDK_DEFAULT_REGION || 'us-east-1';

        const moduleApp = new LakeFormationCdkApp({
          context: {
            ...context,
            module_configs: path.join(__dirname, 'test-config.yaml'),
            additional_accounts: 'xxxxxxxxxxxxx',
            additional_stacks: [
              {
                account: 'xxxxxxxxxxxxx',
                region: region,
              },
            ],
          },
        });
        moduleApp.generateStack();
        return moduleApp;
      },
      {
        module_name: 'test-lakeformation-main',
        org: 'test-org',
        env: 'test-env',
        domain: 'test-domain',
      },
    ),
  );
});
