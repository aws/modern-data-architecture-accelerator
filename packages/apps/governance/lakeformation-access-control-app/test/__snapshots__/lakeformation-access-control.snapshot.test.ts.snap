// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lakeformation-access-control Snapshot Tests Lakeformation App App Snapshot Test 1`] = `
{
  "test-org-test-env-test-domain-test-lakeformation-main": {
    "Description": "(SO0320-lakeformation-access-control) Modern Data Architecture Accelerator. Version 1.1.0",
    "Parameters": {
      "BootstrapVersion": {
        "Default": "/cdk-bootstrap/hnb659fds/version",
        "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
        "Type": "AWS::SSM::Parameter::Value<String>",
      },
      "SsmParameterValuetestorgtestdomaincaefbootstraprolehelperservicetokenC96584B6F00A464EAD1953AFF4B05118Parameter": {
        "Default": "/test-org/test-domain/caef-bootstrap/role-helper-service-token",
        "Type": "AWS::SSM::Parameter::Value<String>",
      },
    },
    "Resources": {
      "RoleResprincipalC": {
        "DeletionPolicy": "Delete",
        "Properties": {
          "ServiceToken": {
            "Ref": "SsmParameterValuetestorgtestdomaincaefbootstraprolehelperservicetokenC96584B6F00A464EAD1953AFF4B05118Parameter",
          },
          "roleRef": {
            "name": "Admin",
            "refId": "principalC",
          },
        },
        "Type": "AWS::CloudFormation::CustomResource",
        "UpdateReplacePolicy": "Delete",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "Properties": {
          "Description": "Stack description parameter to update on version changes",
          "Name": "/test-org/test-domain/test-lakeformation-main/aws-solution",
          "Tags": {
            "mdaa_cdk_app": "lakeformation-access-control",
            "mdaa_domain": "test-domain",
            "mdaa_env": "test-env",
            "mdaa_module_name": "test-lakeformation-main",
            "mdaa_org": "test-org",
          },
          "Type": "String",
          "Value": "(SO0320-lakeformation-access-control) Modern Data Architecture Accelerator. Version 1.1.0",
        },
        "Type": "AWS::SSM::Parameter",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "Properties": {
          "CatalogId": {
            "Ref": "AWS::AccountId",
          },
          "DatabaseInput": {
            "Name": "example_local_resource_link",
            "TargetDatabase": {
              "CatalogId": {
                "Ref": "AWS::AccountId",
              },
              "DatabaseName": "example_database",
            },
          },
        },
        "Type": "AWS::Glue::Database",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "DependsOn": [
          "REPLACED-CFN-LOGICAL-ID",
        ],
        "Properties": {
          "Permissions": [
            "DESCRIBE",
          ],
          "PermissionsWithGrantOption": [],
          "Principal": {
            "DataLakePrincipalIdentifier": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":iam::",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":saml-provider/test-provider:group/test-group1",
                ],
              ],
            },
          },
          "Resource": {
            "Database": {
              "CatalogId": {
                "Ref": "AWS::AccountId",
              },
              "Name": "example_database",
            },
          },
        },
        "Type": "AWS::LakeFormation::PrincipalPermissions",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "DependsOn": [
          "REPLACED-CFN-LOGICAL-ID",
        ],
        "Properties": {
          "Permissions": [
            "DESCRIBE",
          ],
          "PermissionsWithGrantOption": [],
          "Principal": {
            "DataLakePrincipalIdentifier": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":iam::",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":saml-provider/test-provider:user/test-user@example.com",
                ],
              ],
            },
          },
          "Resource": {
            "Database": {
              "CatalogId": {
                "Ref": "AWS::AccountId",
              },
              "Name": "example_database",
            },
          },
        },
        "Type": "AWS::LakeFormation::PrincipalPermissions",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "DependsOn": [
          "REPLACED-CFN-LOGICAL-ID",
        ],
        "Properties": {
          "Permissions": [
            "DESCRIBE",
          ],
          "PermissionsWithGrantOption": [],
          "Principal": {
            "DataLakePrincipalIdentifier": {
              "Fn::GetAtt": [
                "RoleResprincipalC",
                "arn",
              ],
            },
          },
          "Resource": {
            "Database": {
              "CatalogId": {
                "Ref": "AWS::AccountId",
              },
              "Name": "example_database",
            },
          },
        },
        "Type": "AWS::LakeFormation::PrincipalPermissions",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "DependsOn": [
          "REPLACED-CFN-LOGICAL-ID",
        ],
        "Properties": {
          "Permissions": [
            "DESCRIBE",
            "CREATE_TABLE",
            "ALTER",
          ],
          "PermissionsWithGrantOption": [],
          "Principal": {
            "DataLakePrincipalIdentifier": {
              "Fn::GetAtt": [
                "RoleResprincipalC",
                "arn",
              ],
            },
          },
          "Resource": {
            "Database": {
              "CatalogId": {
                "Ref": "AWS::AccountId",
              },
              "Name": "example_database",
            },
          },
        },
        "Type": "AWS::LakeFormation::PrincipalPermissions",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "Properties": {
          "Permissions": [
            "DESCRIBE",
          ],
          "PermissionsWithGrantOption": [],
          "Principal": {
            "DataLakePrincipalIdentifier": {
              "Fn::GetAtt": [
                "RoleResprincipalC",
                "arn",
              ],
            },
          },
          "Resource": {
            "Database": {
              "CatalogId": {
                "Fn::Select": [
                  4,
                  {
                    "Fn::Split": [
                      ":",
                      {
                        "Fn::GetAtt": [
                          "RoleResprincipalC",
                          "arn",
                        ],
                      },
                    ],
                  },
                ],
              },
              "Name": "example_local_resource_link",
            },
          },
        },
        "Type": "AWS::LakeFormation::PrincipalPermissions",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "DependsOn": [
          "REPLACED-CFN-LOGICAL-ID",
        ],
        "Properties": {
          "Permissions": [
            "SELECT",
            "DESCRIBE",
          ],
          "PermissionsWithGrantOption": [],
          "Principal": {
            "DataLakePrincipalIdentifier": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":iam::",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":saml-provider/test-provider:group/test-group1",
                ],
              ],
            },
          },
          "Resource": {
            "Table": {
              "CatalogId": {
                "Ref": "AWS::AccountId",
              },
              "DatabaseName": "example_database",
              "Name": "example_table1",
            },
          },
        },
        "Type": "AWS::LakeFormation::PrincipalPermissions",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "DependsOn": [
          "REPLACED-CFN-LOGICAL-ID",
        ],
        "Properties": {
          "Permissions": [
            "SELECT",
            "DESCRIBE",
          ],
          "PermissionsWithGrantOption": [],
          "Principal": {
            "DataLakePrincipalIdentifier": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":iam::",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":saml-provider/test-provider:user/test-user@example.com",
                ],
              ],
            },
          },
          "Resource": {
            "Table": {
              "CatalogId": {
                "Ref": "AWS::AccountId",
              },
              "DatabaseName": "example_database",
              "Name": "example_table1",
            },
          },
        },
        "Type": "AWS::LakeFormation::PrincipalPermissions",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "DependsOn": [
          "REPLACED-CFN-LOGICAL-ID",
        ],
        "Properties": {
          "Permissions": [
            "SELECT",
            "DESCRIBE",
          ],
          "PermissionsWithGrantOption": [],
          "Principal": {
            "DataLakePrincipalIdentifier": {
              "Fn::GetAtt": [
                "RoleResprincipalC",
                "arn",
              ],
            },
          },
          "Resource": {
            "Table": {
              "CatalogId": {
                "Ref": "AWS::AccountId",
              },
              "DatabaseName": "example_database",
              "Name": "example_table1",
            },
          },
        },
        "Type": "AWS::LakeFormation::PrincipalPermissions",
      },
      "REPLACED-CFN-LOGICAL-ID": {
        "DependsOn": [
          "REPLACED-CFN-LOGICAL-ID",
        ],
        "Properties": {
          "Permissions": [
            "SELECT",
            "DESCRIBE",
            "INSERT",
            "DELETE",
          ],
          "PermissionsWithGrantOption": [],
          "Principal": {
            "DataLakePrincipalIdentifier": {
              "Fn::GetAtt": [
                "RoleResprincipalC",
                "arn",
              ],
            },
          },
          "Resource": {
            "Table": {
              "CatalogId": {
                "Ref": "AWS::AccountId",
              },
              "DatabaseName": "example_database",
              "Name": "example_table2",
            },
          },
        },
        "Type": "AWS::LakeFormation::PrincipalPermissions",
      },
    },
    "Rules": {
      "CheckBootstrapVersion": {
        "Assertions": [
          {
            "Assert": {
              "Fn::Not": [
                {
                  "Fn::Contains": [
                    [
                      "1",
                      "2",
                      "3",
                      "4",
                      "5",
                    ],
                    {
                      "Ref": "BootstrapVersion",
                    },
                  ],
                },
              ],
            },
            "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          },
        ],
      },
    },
  },
  "test-org-test-env-test-domain-test-lakeformation-main-xxxxxxxxxxxxx-us-east-1": {
    "Parameters": {
      "BootstrapVersion": {
        "Default": "/cdk-bootstrap/hnb659fds/version",
        "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
        "Type": "AWS::SSM::Parameter::Value<String>",
      },
    },
    "Rules": {
      "CheckBootstrapVersion": {
        "Assertions": [
          {
            "Assert": {
              "Fn::Not": [
                {
                  "Fn::Contains": [
                    [
                      "1",
                      "2",
                      "3",
                      "4",
                      "5",
                    ],
                    {
                      "Ref": "BootstrapVersion",
                    },
                  ],
                },
              ],
            },
            "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          },
        ],
      },
    },
  },
}
`;

exports[`lakeformation-access-control Snapshot Tests Lakeformation Stack Snapshot Test 1`] = `
{
  "Description": "(SO0320-lakeformation-access-control) Modern Data Architecture Accelerator. Version 1.1.0",
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValuetestorgtestdomaincaefbootstraprolehelperservicetokenC96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/test-org/test-domain/caef-bootstrap/role-helper-service-token",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "RoleResprincipalC": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Ref": "SsmParameterValuetestorgtestdomaincaefbootstraprolehelperservicetokenC96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "roleRef": {
          "name": "Admin",
          "refId": "principalC",
        },
      },
      "Type": "AWS::CloudFormation::CustomResource",
      "UpdateReplacePolicy": "Delete",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "Properties": {
        "Description": "Stack description parameter to update on version changes",
        "Name": "/test-org/test-domain/test-lakeformation-main/aws-solution",
        "Tags": {
          "mdaa_cdk_app": "lakeformation-access-control",
          "mdaa_domain": "test-domain",
          "mdaa_env": "test-env",
          "mdaa_module_name": "test-lakeformation-main",
          "mdaa_org": "test-org",
        },
        "Type": "String",
        "Value": "(SO0320-lakeformation-access-control) Modern Data Architecture Accelerator. Version 1.1.0",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId",
        },
        "DatabaseInput": {
          "Name": "example_local_resource_link",
          "TargetDatabase": {
            "CatalogId": {
              "Ref": "AWS::AccountId",
            },
            "DatabaseName": "example_database",
          },
        },
      },
      "Type": "AWS::Glue::Database",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "DependsOn": [
        "REPLACED-CFN-LOGICAL-ID",
      ],
      "Properties": {
        "Permissions": [
          "DESCRIBE",
        ],
        "PermissionsWithGrantOption": [],
        "Principal": {
          "DataLakePrincipalIdentifier": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::",
                {
                  "Ref": "AWS::AccountId",
                },
                ":saml-provider/test-provider:group/test-group1",
              ],
            ],
          },
        },
        "Resource": {
          "Database": {
            "CatalogId": {
              "Ref": "AWS::AccountId",
            },
            "Name": "example_database",
          },
        },
      },
      "Type": "AWS::LakeFormation::PrincipalPermissions",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "DependsOn": [
        "REPLACED-CFN-LOGICAL-ID",
      ],
      "Properties": {
        "Permissions": [
          "DESCRIBE",
        ],
        "PermissionsWithGrantOption": [],
        "Principal": {
          "DataLakePrincipalIdentifier": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::",
                {
                  "Ref": "AWS::AccountId",
                },
                ":saml-provider/test-provider:user/test-user@example.com",
              ],
            ],
          },
        },
        "Resource": {
          "Database": {
            "CatalogId": {
              "Ref": "AWS::AccountId",
            },
            "Name": "example_database",
          },
        },
      },
      "Type": "AWS::LakeFormation::PrincipalPermissions",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "DependsOn": [
        "REPLACED-CFN-LOGICAL-ID",
      ],
      "Properties": {
        "Permissions": [
          "DESCRIBE",
        ],
        "PermissionsWithGrantOption": [],
        "Principal": {
          "DataLakePrincipalIdentifier": {
            "Fn::GetAtt": [
              "RoleResprincipalC",
              "arn",
            ],
          },
        },
        "Resource": {
          "Database": {
            "CatalogId": {
              "Ref": "AWS::AccountId",
            },
            "Name": "example_database",
          },
        },
      },
      "Type": "AWS::LakeFormation::PrincipalPermissions",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "DependsOn": [
        "REPLACED-CFN-LOGICAL-ID",
      ],
      "Properties": {
        "Permissions": [
          "DESCRIBE",
          "CREATE_TABLE",
          "ALTER",
        ],
        "PermissionsWithGrantOption": [],
        "Principal": {
          "DataLakePrincipalIdentifier": {
            "Fn::GetAtt": [
              "RoleResprincipalC",
              "arn",
            ],
          },
        },
        "Resource": {
          "Database": {
            "CatalogId": {
              "Ref": "AWS::AccountId",
            },
            "Name": "example_database",
          },
        },
      },
      "Type": "AWS::LakeFormation::PrincipalPermissions",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "Properties": {
        "Permissions": [
          "DESCRIBE",
        ],
        "PermissionsWithGrantOption": [],
        "Principal": {
          "DataLakePrincipalIdentifier": {
            "Fn::GetAtt": [
              "RoleResprincipalC",
              "arn",
            ],
          },
        },
        "Resource": {
          "Database": {
            "CatalogId": {
              "Fn::Select": [
                4,
                {
                  "Fn::Split": [
                    ":",
                    {
                      "Fn::GetAtt": [
                        "RoleResprincipalC",
                        "arn",
                      ],
                    },
                  ],
                },
              ],
            },
            "Name": "example_local_resource_link",
          },
        },
      },
      "Type": "AWS::LakeFormation::PrincipalPermissions",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "DependsOn": [
        "REPLACED-CFN-LOGICAL-ID",
      ],
      "Properties": {
        "Permissions": [
          "SELECT",
          "DESCRIBE",
        ],
        "PermissionsWithGrantOption": [],
        "Principal": {
          "DataLakePrincipalIdentifier": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::",
                {
                  "Ref": "AWS::AccountId",
                },
                ":saml-provider/test-provider:group/test-group1",
              ],
            ],
          },
        },
        "Resource": {
          "Table": {
            "CatalogId": {
              "Ref": "AWS::AccountId",
            },
            "DatabaseName": "example_database",
            "Name": "example_table1",
          },
        },
      },
      "Type": "AWS::LakeFormation::PrincipalPermissions",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "DependsOn": [
        "REPLACED-CFN-LOGICAL-ID",
      ],
      "Properties": {
        "Permissions": [
          "SELECT",
          "DESCRIBE",
        ],
        "PermissionsWithGrantOption": [],
        "Principal": {
          "DataLakePrincipalIdentifier": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::",
                {
                  "Ref": "AWS::AccountId",
                },
                ":saml-provider/test-provider:user/test-user@example.com",
              ],
            ],
          },
        },
        "Resource": {
          "Table": {
            "CatalogId": {
              "Ref": "AWS::AccountId",
            },
            "DatabaseName": "example_database",
            "Name": "example_table1",
          },
        },
      },
      "Type": "AWS::LakeFormation::PrincipalPermissions",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "DependsOn": [
        "REPLACED-CFN-LOGICAL-ID",
      ],
      "Properties": {
        "Permissions": [
          "SELECT",
          "DESCRIBE",
        ],
        "PermissionsWithGrantOption": [],
        "Principal": {
          "DataLakePrincipalIdentifier": {
            "Fn::GetAtt": [
              "RoleResprincipalC",
              "arn",
            ],
          },
        },
        "Resource": {
          "Table": {
            "CatalogId": {
              "Ref": "AWS::AccountId",
            },
            "DatabaseName": "example_database",
            "Name": "example_table1",
          },
        },
      },
      "Type": "AWS::LakeFormation::PrincipalPermissions",
    },
    "REPLACED-CFN-LOGICAL-ID": {
      "DependsOn": [
        "REPLACED-CFN-LOGICAL-ID",
      ],
      "Properties": {
        "Permissions": [
          "SELECT",
          "DESCRIBE",
          "INSERT",
          "DELETE",
        ],
        "PermissionsWithGrantOption": [],
        "Principal": {
          "DataLakePrincipalIdentifier": {
            "Fn::GetAtt": [
              "RoleResprincipalC",
              "arn",
            ],
          },
        },
        "Resource": {
          "Table": {
            "CatalogId": {
              "Ref": "AWS::AccountId",
            },
            "DatabaseName": "example_database",
            "Name": "example_table2",
          },
        },
      },
      "Type": "AWS::LakeFormation::PrincipalPermissions",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
